### ISSUE TITLE
`history-token-search-backward` erases whole line on `main` branch

### ISSUE THREAD

**Issue by zx8** [fish-shell/fish-shell#10145]
Using https://github.com/fish-shell/fish-shell/commit/f49efcc691781531b9f927b3e4d0a306d4e8f510

---

First, build up some history:

```bash
$ echo github.com/fish-shell/fish-shell
github.com/fish-shell/fish-shell

$ echo https://github.com/fish-shell/fish-shell/commit/f49efcc691781531b9f927b3e4d0a306d4e8f510
https://github.com/fish-shell/fish-shell/commit/f49efcc691781531b9f927b3e4d0a306d4e8f510
```

Now try to recall the first token containing "fish":

```
$ echo fish<history-token-search-backward>
```

This works as expected and the commandline is updated with the suggestion:

```
$ echo https://github.com/fish-shell/fish-shell/commit/f49efcc691781531b9f927b3e4d0a306d4e8f510
```

Now, try again:

```
$ echo fish<history-token-search-backward>
```

This replaces the entire commandline with the suggestion:

```
$ github.com/fish-shell/fish-shell
```

I would expect the commandline to be:

```
$ echo github.com/fish-shell/fish-shell
```

Interestingly, I cannot reproduce this using `fish --no-config` but can reproduce with `sh -c 'env HOME=$(mktemp -d) fish`

**Comment by zx8** [comment]
Not sure if related, but `undo` (ctrl+z) seems to erase the whole line in some circumstances (trying to reproduce at the moment):

```
$ echo foo<ctrl+z>
```

Results in:

```
$
```

Since this issue is similar to the OP, I'll wait for someone to triage and confirm whether or not they are related before opening up a separate issue.

**Comment by krobelus** [comment]
I could reproduce only once. This sounds like something ASan would detect.
Can you build with something like

```
rustup default nightly
export RUSTFLAGS=-Zsanitizer=address
export CXXFLAGS="-g -O1 -fno-omit-frame-pointer -fsanitize=undefined -fsanitize=address -fsanitize-blacklist=$PWD/build_tools/asan_blacklist.txt"
export ASAN_OPTIONS=check_initialization_order=1:detect_stack_use_after_return=1:detect_leaks=1:fast_unwind_on_malloc=0
export UBSAN_OPTIONS=print_stacktrace=1:report_error_type=1
export LSAN_OPTIONS=verbosity=0:log_threads=0:use_tls=1:print_suppressions=0:suppressions=$PWD/build_tools/lsan_suppressions.txt
rm -rf build
mkdir build
cd build
cmake .. -GNinja -DCMAKE_BUILD_TYPE=Debug -DASAN=1 -DRust_CARGO_TARGET=x86_64-unknown-linux-gnu' &&
ninja fish
```

(adjust `Rust_CARGO_TARGET` depending on platform)

**Comment by krobelus** [comment]
nevermind I can reproduce consistently
